src/Orleans.Core/Runtime/ExecutorService.cs

```diff
@@ -1,15 +1,54 @@
-﻿using System.Threading.Tasks;
-using Orleans.Runtime.Scheduler;
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using Orleans.Messaging;
 
 namespace Orleans.Runtime
 {
-    internal class ExecutorService : ITaskScheduler
+    internal interface IExecutor
     {
-        private readonly TaskScheduler taskScheduler = new ThreadPerTaskScheduler(task => (task as AsynchAgentTask)?.Name);
+        void QueueWorkItem(WaitCallback callback, object state = null);
 
-        public void RunTask(Task task)
+        int WorkQueueCount { get; }
+    }
+
+    internal class ExecutorService
+    {
+        public IExecutor GetExecutor(GetExecutorRequest request)
+        {
+            if (typeof(GatewayConnection).IsAssignableFrom(request.StageType))
```

Comment:
I find this a bit strange, as though `GetExecutor` has to know too much about its caller. Could there be a different way to identify the required kind of `IExecutor` which still separates policy & mechanism?

Reply:
If to make completely reusable component - that could be a problem. But I think for this case, where list of stages is known beforehand and almost constant - this knowledge will allow greater flexibility in assigning executors to stages. Something alike was implemented in ElasticSearch - 
https://github.com/elastic/elasticsearch/blob/8caf7d4ff8738131ae65cca86869c904d0e5f19b/core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java#L123